### ISSUE TITLE
-[ARTBaseMessage id] should not be nullable

### ISSUE THREAD

**Issue by lawrence-forooghian** [ably/ably-cocoa#1200]
## Summary
`ARTBaseMessage`, an abstract class which is the superclass of `ARTMessage` and `ARTPresenceMessage`, has a nullable `id` property in its interface.

Except in a couple of very localised places, this value will in practice never be `nil`. But the nullability of the property in the interface necessitates `nil`-handling behaviour in far-flung parts of the codebase. The `nil`-case behaviour will, in practice, never be triggered and hence we have some dead code.

We should refine the interface to make it clear that this property will not usually be `nil`, and we can then remove the dead code.

### Why will this value "in practice never be `nil`"?
See [`TM2a`](https://ably.com/documentation/client-lib-development-guide/features#TM2a) and [`TP3a`](https://ably.com/documentation/client-lib-development-guide/features#TP3a): in the case where the client receives a message without an `id`, it will synthesize a value for this property (see implementations for [`ARTMessage`](https://github.com/ably/ably-cocoa/blob/51de8ebd5b4f40427405f69f9d9cc9e81228decc/Source/ARTRealtimeChannel.m#L847-L849) and [`ARTPresenceMessage`](https://github.com/ably/ably-cocoa/blob/51de8ebd5b4f40427405f69f9d9cc9e81228decc/Source/ARTRealtimeChannel.m#L880-L882)).

### What are the "couple of very localised places"?
The `ARTRealtime` methods I've linked to above, which take a message [DTO](https://en.wikipedia.org/wiki/Data_transfer_object) received from Realtime and populate the `id` property if necessary.

### What's an example of the dead code?
[This null check in `-[ARTPresenceMessage parseId]`](https://github.com/ably/ably-cocoa/blob/51de8ebd5b4f40427405f69f9d9cc9e81228decc/Source/ARTPresenceMessage.m#L51-L53).

### How might we solve it?
I think the problem is that we’re using a single type to represent a few different things which are similar but not identical. Let's take `ARTPresenceMessage` as an example. This type is currently used to represent:

1. A DTO received from Realtime, which may have a `nil` ID. This object is localised to the aforementioned internal `ARTRealtime` methods, whose job is to convert this object into an object of type 2 below.
1. The "presence message" object that's passed around the codebase. This will always have a non-`nil` ID.
1. A DTO for sending to Realtime when we publish a presence message, which has a `nil` ID. This object is localised to the internal [`-[ARTRealtimePresence enterOrUpdateAfterChecks:clientId:data:callback:]`](https://github.com/ably/ably-cocoa/blob/51de8ebd5b4f40427405f69f9d9cc9e81228decc/Source/ARTRealtimePresence.m#L326) and [`-[ARTRealtimePresence leaveAfterChecks:data:callback:]`](https://github.com/ably/ably-cocoa/blob/51de8ebd5b4f40427405f69f9d9cc9e81228decc/Source/ARTRealtimePresence.m#L382) methods.

I think we should represent these as different types (types 1 and 3 might actually be the same thing).

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-4289) by [Unito](https://www.unito.io)

**Comment by QuintinWillison** [comment]
I've had a brief read of your opening comment on this issue, @lawrence-forooghian. It all sounds very logical and I completely trust your analysis.

It did immediately remind me of previous discussions had around our asset tracking product - see https://github.com/ably/ably-asset-tracking-android/issues/165.

This kind of improvement in the codebase will always be a good idea, as I imagine it will improve developer experience for maintainers. It does make me wonder whether there might be a similar opportunity in other SDK codebases, which would lead us to consider splitting DTOs ('the Ably protocol level bits') to their own area of the spec.. I've always struggled with the level of conflation of concerns in our current spec so this could be an achievable leap forward in terms of pulling that apart a bit. 🤔

**Comment by lawrence-forooghian** [comment]
Yeah, I was wondering if there already existed some separate documentation of the DTOs, perhaps in some place that I haven't seen / don't have access to (i.e. documentation for the API / Realtime services)?

**Comment by maratal** [comment]
This will help to track nullability issues #1213
